Support pyenv on Windows

nre-ableton opened the pull request:
This PR adds support for Windows in `pyenv.createVirtualEnv()`. I've bumped the minor version since this is technically a breaking change, as these functions previously resulted in an error being thrown when run on Windows.

nre-ableton commented on src/com/ableton/VirtualEnv.groovy:
```diff
@@ -37,24 +37,29 @@ class VirtualEnv implements Serializable {
+    // Cache for later usage
+    boolean isUnix = script.isUnix()
+    String safePyenvRoot = isUnix ? pyenvRoot : convertWindowsPath(pyenvRoot)
+    List activateCommands = ["export PATH=${safePyenvRoot}/bin:\$PATH"]
```
I'm not sure I see it. Maybe I'm missing something obvious? Either way, would it be ok to squash so that there aren't so many fixups cluttering the review?

nre-ableton replied:
Ok, now that I've squashed and looked at the commits in the branch, I still don't understand this comment. Can you re-review and clarify please?

ala-ableton replied:
In 31ed1db0d80e8133d34fa78d5a06cca5cffa475b, the old version had `export PATH=...`, the new version doesn't,

nre-ableton replied:
Ah, ok, now I see it. Please see 7812313.

nre-ableton commented:
> In 43d3b35720affb2c67dcf1a97a2e00ba4ef10fa7, we now have `export PATH=...` twice.

Sorry about that. Fixed in b5b1c96bd326dd56133e2b8a21364665d93e8d3a.

ala-ableton commented on src/com/ableton/VirtualEnv.groovy:
```diff
@@ -37,34 +37,41 @@ class VirtualEnv implements Serializable {
+    List activateCommands = ["export PATH=${safePyenvRoot}/bin:\$PATH"]
+    if (!script.env.PYENV_ROOT) {
+      // Only export PYENV_ROOT if it isn't part of the Jenkins environment
+      activateCommands.add("export PYENV_ROOT=${pyenvRoot}")
```
Why not use `safePyenvRoot` here?

ala-ableton reviewed (changes requested):
Please rebase/squash, then re-rebase and edit 43d3b35720affb2c67dcf1a97a2e00ba4ef10fa7 to make sure that `export PYENV_ROOT=...`  keeps on coming before `export PATH=...`, instead of having `export PATH=...` move around.

ala-ableton commented on src/com/ableton/VirtualEnv.groovy:
```diff
@@ -48,12 +48,13 @@ class VirtualEnv implements Serializable {
+    List activateCommands = ['export PATH=\$PYENV_ROOT/bin:\$PATH']
+    if (!script.env.PYENV_ROOT) {
+      // Only export PYENV_ROOT if it isn't part of the Jenkins environment
+      activateCommands.add("export PYENV_ROOT=${pyenvRoot}")
```
In a4881678645f5b8c3d2c2019cae84850eea7dcdf, please don't put `export PATH=...` before `export PYENV_ROOT=...`

nre-ableton replied:
I'll take care of this when I'm back from holidays. Also, I merged https://github.com/Ableton/python-pipeline-utils/pull/95 first, and I will rework this PR slightly to take advantage of that.

nre-ableton commented:
So I've rebased this PR, and also added Windows support for the newly-added integration tests, but they keep timing out after 1 hour.

The first time, the build failed with a different error (not a timeout):

```
[2022-07-28T08:33:57.000Z] F:/jenkins-node/workspace@tmp/durable-6cb19c97/script.sh: line 4: /f/jenkins-node/workspace/.venv/venv-95957963/Scripts/activate: No such file or directory
```

I've no idea why this is happening. :disappointed:

nre-ableton commented:
Ok! This PR is finally ready for review, since our CI nodes now have the updated version of Pyenv needed for Windows support.

ala-ableton commented on vars/pyenv.groovy:
```diff
@@ -10,7 +10,9 @@ import com.ableton.VirtualEnv
 VirtualEnv createVirtualEnv(String pythonVersion, String pyenvRoot = null) {
-  return new Pyenv(this, pyenvRoot ?: env.PYENV_ROOT).createVirtualEnv(pythonVersion)
+  return new Pyenv(
+    this, VirtualEnv.posixPath(pyenvRoot ?: env.PYENV_ROOT)
```
Why do 2 conversions? One here and one in `Pyenv.createVirtualEnv`? Wouldn't it be simpler to do a single conversion in the constructor of the `Pyenv` object?

nre-ableton replied:
I was trying to avoid doing too much work in the ctor. But I agree it is probably better in this case compared to the code duplication. Fixed in 735ff99.

ala-ableton commented on src/com/ableton/Pyenv.groovy:
```diff
@@ -120,8 +120,8 @@ class Pyenv implements Serializable {
 
-    if (!script.fileExists(pyenvRoot)) {
-      script.error "pyenv root path '${pyenvRoot}' does not exist"
+    if (script.sh(returnStatus: true, script: "${pyenvRoot}/bin/pyenv --version")) {
```
So `script.dir` is fine with POSIX paths on Windows, but not `script.fileExists`... That's strange 🤔

nre-ableton replied:
No, neither one of them works. The problem is with *absolute, CygWin-style paths*. I just ran this code in the sandbox:

```groovy
node('generic-win') {
    dir('C:\\Windows\\Temp') {
        bat 'dir'
    }
    if (fileExists('C:\\Windows\\Temp\\vmware-vmusr-ci.log')) {
        echo 'it is there'
    } else {
        echo 'it is not there'
    }
    
    dir('/c/windows/temp') {
        bat 'dir'
    }
    if (fileExists('/c/windows/temp/vmware-vmusr-ci.log')) {
        echo 'it is also there'
    } else {
        echo 'it is also not there'
    }
}
```

Here's the result:

```
Running on [generic-win-vs2019-00](https://newci.office.ableton.com/computer/generic%2Dwin%2Dvs2019%2D00/) in F:\workspace\Utilities\sandbox
 [Pipeline] {
 [Pipeline] dir
 Running in C:\Windows\Temp
 [Pipeline] {
 [Pipeline] bat
 
 C:\Windows\Temp>dir
  Volume in drive C has no label.
  Volume Serial Number is 4CEB-A7DB
 
  Directory of C:\Windows\Temp
 
 22/09/2023  13:29    <DIR>          .
 22/09/2023  13:29    <DIR>          ..
 17/09/2023  07:46         3.788.840 f03139a7-0df9-4c0e-8fae-a95623bd1935.tmp
 27/07/2023  13:23                 0 FXSAPIDebugLogFile.txt
 27/07/2023  13:23                 0 FXSTIFFDebugLogFile.txt
 22/09/2023  13:31           931.062 MpCmdRun.log
 22/09/2023  12:08           957.232 MpSigStub.log
 22/09/2023  09:46           517.550 msedge_installer.log
 20/09/2023  09:51    <DIR>          vmware-SYSTEM
 14/09/2023  23:19           345.309 vmware-vmsvc-SYSTEM.log
 26/07/2023  11:29            44.173 vmware-vmusr-ci.log
 14/09/2023  21:33    <DIR>          WinSAT
 14/09/2023  14:34    <DIR>          {5ED6833A-C365-4743-9B64-12A039EC1E06}
                8 File(s)      6.584.166 bytes
                5 Dir(s)  15.565.709.312 bytes free
 [Pipeline] }
 [Pipeline] // dir
 [Pipeline] fileExists
 [Pipeline] echo
 it is there
 [Pipeline] dir
 Running in /c/windows/temp
 [Pipeline] {
 [Pipeline] bat
 
 F:\c\windows\temp>dir
  Volume in drive F has no label.
  Volume Serial Number is 0C9E-D458
 
  Directory of F:\c\windows\temp
 
 22/09/2023  14:49    <DIR>          .
 22/09/2023  14:49    <DIR>          ..
                0 File(s)              0 bytes
                2 Dir(s)  20.222.361.600 bytes free
 [Pipeline] }
 [Pipeline] // dir
 [Pipeline] fileExists
 [Pipeline] echo
 it is also not there
```

ala-ableton replied:
> neither one of them works.

Then this means that https://github.com/Ableton/python-pipeline-utils/pull/94/commits/59e794c124a1aff0c263c95c9255d5ef97509bc2 breaks running `VirtualEnv.cleanup` on Windows.

nre-ableton replied:
Good catch. Fixed in 37fc6a4.

ala-ableton commented on src/com/ableton/Pyenv.groovy:
```diff
@@ -51,10 +47,14 @@ class Pyenv implements Serializable {
+      installCommands.add("export PATH=${VirtualEnv.posixPath(pyenvRoot)}/bin:\$PATH")
+      if (script.isUnix()) {
+        installCommands += [
+          "eval \"\$(pyenv init --path)\"",
```
`eval "$(pyenv init --path)"` prepends `$PYENV_ROOT/shims` to the `PATH` env var. Isn't there a `shims` folder on Windows?

nre-ableton replied:
Looking at the production node that ran this CI job, it seems so. It's in `F:\pyenv-win\pyenv-win\shims`. Fixed in 24f6bf9.

nre-ableton commented on src/com/ableton/Pyenv.groovy:
```diff
@@ -46,8 +46,12 @@ class Pyenv implements Serializable {
+      List installCommands = []
+      if (!script.env.PYENV_ROOT) {
+        // Only export PYENV_ROOT if it isn't part of the Jenkins environment
+        installCommands.add("export PYENV_ROOT=${pyenvRoot}")
```
I'm not sure I follow you here. Why do you say, "if it might get ignored"? If `pyenvRoot` is added as the optional argument to either function in the `pyenv` singleton, it'll be set here, not ignored.

The purpose of this line is to just prevent exporting something which is already in the environment.

ala-ableton replied:
The `pyenvRoot` field won't be used here if `script.env.PYENV_ROOT` is true-ish. That's what I mean by "ignored".

nre-ableton replied:
I think that the idea here was that we didn't want to have "conflicting" values, but come to think of it, maybe we _do_ always want to export `PYENV_ROOT` because we have determined the "safe" value. Fixed in 2d0d2c7 (which means we'll drop that commit). I'll do some tests as well.

nre-ableton commented on src/com/ableton/Pyenv.groovy:
```diff
@@ -52,7 +52,7 @@ class Pyenv implements Serializable {
       }
       installCommands += [
-        "export PATH=\$PYENV_ROOT/bin:\$PATH",
+        "export PATH=${VirtualEnv.posixPath(pyenvRoot)}/bin:\$PATH",
```
I guess we should do both. Fixed in 34f1963.

nre-ableton commented on src/com/ableton/VirtualEnv.groovy:
```diff
@@ -70,6 +70,14 @@ class VirtualEnv implements Serializable {
+    if (path[1] == ':') {
+      return "/${path[0].toLowerCase()}/${path.substring(3).replaceAll('\\\\', '/')}"
+    }
+    return path
```
We can easily support relative paths. Fixed in 65a8eea.

nre-ableton commented on src/com/ableton/Pyenv.groovy:
```diff
@@ -39,25 +39,32 @@ class Pyenv implements Serializable {
+    int result
+    script.withEnv(["PYENV_VERSION=${pythonVersion}"]) {
+      List installCommands = []
+      String safePyenvRoot = VirtualEnv.posixPath(pyenvRoot)
```
But it's not "safe" in the existing environment. For instance, on our CI nodes, it is exported in the environment to a Windows path.

ala-ableton replied:
I'm talking about the `pyenvRoot` field. Right here. `this.pyenvRoot` if you prefer.

nre-ableton replied:
Fixed in 5d4b875.

nre-ableton commented:
@ala-ableton Please let me know when I should rebase.

ala-ableton commented:
@nre-ableton Feel free to rebase whenever you want.

nre-ableton commented:
> @nre-ableton Feel free to rebase whenever you want.

Done. Feel free to review. 😊 Note that I pushed a dropme commit just to make sure that installing fresh (uncached) Python versions still works.

ala-ableton commented on src/com/ableton/Pyenv.groovy:
```diff
@@ -44,19 +44,21 @@ class Pyenv implements Serializable {
+      'pip install virtualenv',
+      "virtualenv ${venv.venvRootDir}",
+    ]
+    result = venv.script.sh(
```
In de32499621186c7c7d63e2bac97f5c8bcf46e265, I don't see why `result` gets defined before `installCommands`.

nre-ableton replied:
See 14fbba1.

ala-ableton commented on src/com/ableton/Pyenv.groovy:
```diff
@@ -45,21 +45,22 @@ class Pyenv implements Serializable {
-      "eval \"\$(pyenv init --path)\"",
-      "eval \"\$(pyenv init -)\"",
-      "pyenv install --skip-existing ${pythonVersion}",
-      "pyenv shell ${pythonVersion}",
```
pyenv-win supports `pyenv shell`, so I'm not sure why we can't use it on Windows.

nre-ableton replied:
IIRC, when I last tested this, it didn't work for us. I'll try it again, though.

nre-ableton replied:
I just pushed a WIP commit to test this, and indeed, it failed. So while it is supported, it doesn't seem to work from Jenkins for the shell session.

ala-ableton replied:
That's a shame... Well, I guess setting the `PYTHON_VERSION` env var is not that bad.

ala-ableton commented on src/com/ableton/Pyenv.groovy:
```diff
@@ -48,7 +48,7 @@ class Pyenv implements Serializable {
       List installCommands = [
         "export PYENV_ROOT=${pyenvRoot}",
-        "export PATH=\$PYENV_ROOT/bin:\$PATH",
+        "export PATH=${pyenvRoot}/bin:\$PATH",
```
In 2a9caf3ab781cbc9575271a6a96b65314ebcd2a3, I don't see why we need to change this line.

nre-ableton replied:
See 4270827.

nre-ableton commented:
Forgive the rebase/force-push, I just noticed a vim-related typo in a commit message and decided to just drop the experiment commits regarding `pyenv shell`.

nre-ableton commented:
@ala-ableton I've rebased this PR, ptal!

ala-ableton commented on src/com/ableton/VirtualEnv.groovy:
```diff
@@ -70,6 +70,14 @@ class VirtualEnv implements Serializable {
     this.venvBinDir = "${venvRootDir}/${activateSubDir}"
   }
 
+  @NonCPS
+  static final String posixPath(String path) {
```
Does it really need to be annotated with `@NonCPS`?

nre-ableton replied:
Probably not, but why not? It is non-CPS-compatible, and we already need to import that symbol for `randomName`.

ala-ableton replied:
I guess you misunderstood my question.
```suggestion
  static final String posixPath(String path) {
```

nre-ableton replied:
No, I understood your question correctly. As the function does not use any pipeline steps, it should ideally by annotated with `@NonCPS` in order to optimize Jenkins' performance.

ala-ableton commented on src/com/ableton/Pyenv.groovy:
```diff
@@ -23,7 +23,7 @@ class Pyenv implements Serializable {
   Pyenv(Object script, String pyenvRoot) {
     this.script = script
-    this.pyenvRoot = pyenvRoot
+    this.pyenvRoot = pyenvRoot ? VirtualEnv.posixPath(pyenvRoot) : null
```
Fixup for f93cbac65f60334e86e93d5a959ef386103fbd09 (whose commit message should also be reworded)
```suggestion
    this.pyenvRoot =
      script.env.OS == 'Windows_NT' ? pyenvRoot.replace('\\', '/') : pyenvRoot
```
It might actually make sense to simplify squash that change into "Support pyenv on Windows".

ala-ableton reviewed (changes requested):
Please rebase on the latest `main` state and drop the following commits:
- 4b31949270f42e88eed70fa5af4a5992d0761b2c
- 994338699016c75246c9f5ca3360d7a1fb84057d
- ab8dc4d5be8d680f9f1611fb0ecb591d004ac04c

nre-ableton commented:
I've rebased and edited the history as requested, but there are some failing tests that I will investigate next week. I will re-request a review once these have been fixed.

ala-ableton commented on src/com/ableton/Pyenv.groovy:
```diff
@@ -23,7 +23,7 @@ class Pyenv implements Serializable {
   Pyenv(Object script, String pyenvRoot) {
     this.script = script
-    this.pyenvRoot = pyenvRoot
+    this.pyenvRoot = pyenvRoot ? VirtualEnv.posixPath(pyenvRoot) : null
```
Fixup for b82526ed0a66ae10ef963fd10d196f7c0f5ddf5e
```suggestion
    this.pyenvRoot =
      script.env.OS == 'Windows_NT' ? pyenvRoot.replace('\\', '/') : pyenvRoot
```

nre-ableton replied:
Yeah, this won't work, we'll just get paths like `c:/mock/pyenv`. We still need 4b31949, or at least the equivalent logic.

nre-ableton replied:
I don't think this is going to work, and we probably need to re-cherry-pick https://github.com/Ableton/python-pipeline-utils/commit/4b31949270f42e88eed70fa5af4a5992d0761b2c (and friends). Although the unit tests pass with their mocks, it fails when running on a Windows host with a `pyenv: command not found` error. Have a look at the `PATH` that we create:

> `PATH='F:/pyenv-win/pyenv-win/shims:F:/pyenv-win/pyenv-win/bin:/c/Python27:/c/Python27/Scripts:/c/Program Files/Eclipse Adoptium/jdk-21.0.3.9-hotspot/bin:/c/WINDOWS/system32:/c/WINDOWS:/c/WINDOWS/System32/Wbem:/c/WINDOWS/System32/WindowsPowerShell/v1.0:/c/ProgramData/chocolatey/bin:/cmd:/mingw64/bin:/usr/bin:/c/Program Files/LLVM/bin:/c/Program Files/Git LFS:/c/Python37:/c/Python37/Scripts:/c/Python310:/c/Python310/Scripts:/c/Program Files/CMake/bin:/c/Users/ci/AppData/Local/Microsoft/WindowsApps'`

I'm pretty sure that these path elements need to be cygwin-compatible paths like `/c/Windows/whatever`.

ala-ableton commented on test/com/ableton/PyenvTest.groovy:
```diff
@@ -101,9 +101,20 @@ class PyenvTest extends BasePipelineTest {
+    helper.addShMock("${cygwinPyenvRoot}/bin/pyenv install --list", '''Available versions:
+  1.2.3
+''', 0)
+    helper.addShMock(installCommands(cygwinPyenvRoot, pythonVersion, false), '', 0)
```
That mock is not called.

ala-ableton reviewed (changes requested):
Please apply changes similar to a4c566d9a06c5b2aa464fc0e9e4435b0ac735d94 to Windows-only tests.

nre-ableton commented:
@ala-ableton I’m ok with your fixup, so I’ll go ahead and squash

ala-ableton reviewed (approved):
LGTM 👍🏻
